File: java/src/SegmentTree.java
import java.util.*;

public class SegmentTree {

	// Modify the following 5 methods to implement your custom operations on the tree.
	// This example implements Add/Max operations. Operations like Add/Sum, Set/Max can also be implemented.
	int modifyOperation(int x, int y) {
		// treat neutral delta separately
		if (x == getNeutralDelta()) {
			return y;
		}
		if (y == getNeutralDelta()) {
			return x;
		}
		// define operation here
		return x + y;
	}

	int queryOperation(int x, int y) {
		return Math.max(x, y);
	}

	int deltaEffectOnSegment(int delta, int segmentLength) {
		// Here you must write a fast equivalent of following slow code:
		// int result = delta;
		// for (int i = 0; i < segmentLength - 1; i++) result = queryOperation(result, delta);
		// return result;
		return delta;
	}

	int getNeutralDelta() {
		return Integer.MIN_VALUE;
	}

	int getInitValue() {
		return 0;
	}

	// generic code
	int n;
	int[] value;
	int[] delta; // delta[i] affects value[i], delta[2*i+1] and delta[2*i+2]

	void pushDelta(int root, int left, int right) {
		value[root] = modifyOperation(value[root], deltaEffectOnSegment(delta[root], right - left + 1));
		delta[2 * root + 1] = modifyOperation(delta[2 * root + 1], delta[root]);
		delta[2 * root + 2] = modifyOperation(delta[2 * root + 2], delta[root]);
		delta[root] = getNeutralDelta();
	}

	public SegmentTree(int n) {
		this.n = n;
		value = new int[4 * n];
		delta = new int[4 * n];
		init(0, 0, n - 1);
	}

	void init(int root, int left, int right) {
		if (left == right) {
			value[root] = getInitValue();
			delta[root] = getNeutralDelta();
		} else {
			int mid = (left + right) >> 1;
			init(2 * root + 1, left, mid);
			init(2 * root + 2, mid + 1, right);
			value[root] = queryOperation(value[2 * root + 1], value[2 * root + 2]);
			delta[root] = getNeutralDelta();
		}
	}

	public int query(int a, int b) {
		return query(a, b, 0, 0, n - 1);
	}

	int query(int a, int b, int root, int left, int right) {
		if (a == left && b == right)
			return modifyOperation(value[root], deltaEffectOnSegment(delta[root], right - left + 1));
		pushDelta(root, left, right);
		int mid = (left + right) >> 1;
		if (a <= mid && b > mid)
			return queryOperation(
					query(a, Math.min(b, mid), root * 2 + 1, left, mid),
					query(Math.max(a, mid + 1), b, root * 2 + 2, mid + 1, right));
		else if (a <= mid)
			return query(a, Math.min(b, mid), root * 2 + 1, left, mid);
		else if (b > mid)
			return query(Math.max(a, mid + 1), b, root * 2 + 2, mid + 1, right);
		throw new RuntimeException("Incorrect query from " + a + " to " + b);
	}

	public void modify(int a, int b, int delta) {
		modify(a, b, delta, 0, 0, n - 1);
	}

	void modify(int a, int b, int delta, int root, int left, int right) {
		if (a == left && b == right) {
			this.delta[root] = modifyOperation(this.delta[root], delta);
			return;
		}
		pushDelta(root, left, right);
		int mid = (left + right) >> 1;
		if (a <= mid)
			modify(a, Math.min(b, mid), delta, 2 * root + 1, left, mid);
		if (b > mid)
			modify(Math.max(a, mid + 1), b, delta, 2 * root + 2, mid + 1, right);
		value[root] = queryOperation(
				modifyOperation(value[2 * root + 1], deltaEffectOnSegment(this.delta[2 * root + 1], mid - left + 1)),
				modifyOperation(value[2 * root + 2], deltaEffectOnSegment(this.delta[2 * root + 2], right - mid)));
	}

	// Random test
	public static void main(String[] args) {
		Random rnd = new Random();
		for (int step = 0; step < 1000; step++) {
			int n = rnd.nextInt(50) + 1;
			int[] x = new int[n];
			SegmentTree t = new SegmentTree(n);
			Arrays.fill(x, t.getInitValue());
			for (int i = 0; i < 1000; i++) {
				int b = rnd.nextInt(n);
				int a = rnd.nextInt(b + 1);
				int cmd = rnd.nextInt(3);
				if (cmd == 0) {
					int delta = rnd.nextInt(100) - 50;
					t.modify(a, b, delta);
					for (int j = a; j <= b; j++)
						x[j] = t.modifyOperation(x[j], delta);
				} else if (cmd == 1) {
					int res1 = t.query(a, b);
					int res2 = x[a];
					for (int j = a + 1; j <= b; j++)
						res2 = t.queryOperation(res2, x[j]);
					if (res1 != res2)
						throw new RuntimeException("error");
				} else {
					for (int j = 0; j < n; j++) {
						if (t.query(j, j) != x[j])
							throw new RuntimeException("error");
					}
				}
			}
		}
		System.out.println("Test passed");
	}
}

File: scala/src/SegmentTree.scala
import java.util.Random

class SegmentTree(n: Int) {
  // Modify the following 5 methods to implement your custom operations on the tree.
  // This example implements Add/Max operations. Operations like Add/Sum, Set/Max can also be implemented.
  def modifyOperation(x: Int, y: Int) = x + y

  // query (or combine) operation
  def queryOperation(leftValue: Int, rightValue: Int): Int = math.max(leftValue, rightValue)

  def deltaEffectOnSegment(delta: Int, segmentLength: Int): Int = {
    if (delta == getNeutralDelta) return getNeutralDelta
    // Here you must write a fast equivalent of following slow code:
    // int result = delta;
    // for (int i = 1; i < segmentLength; i++) result = queryOperation(result, delta);
    // return result;
    delta
  }

  def getNeutralDelta = 0

  def getInitValue = 0

  val value = new Array[Int](4 * n)
  val delta = new Array[Int](4 * n) // delta[i] affects value[i], delta[2*i+1] and delta[2*i+2]

  init(0, 0, n - 1)

  def init(root: Int, left: Int, right: Int) {
    if (left == right) {
      value(root) = getInitValue
      delta(root) = getNeutralDelta
    } else {
      val mid = (left + right) >> 1
      init(2 * root + 1, left, mid)
      init(2 * root + 2, mid + 1, right)
      value(root) = queryOperation(value(2 * root + 1), value(2 * root + 2))
      delta(root) = getNeutralDelta
    }
  }

  def joinValueWithDelta(value: Int, delta: Int): Int = {
    if (delta == getNeutralDelta) return value
    modifyOperation(value, delta)
  }

  def joinDeltas(delta1: Int, delta2: Int): Int = {
    if (delta1 == getNeutralDelta) return delta2
    if (delta2 == getNeutralDelta) return delta1
    modifyOperation(delta1, delta2)
  }

  def pushDelta(root: Int, left: Int, right: Int) {
    value(root) = joinValueWithDelta(value(root), deltaEffectOnSegment(delta(root), right - left + 1))
    delta(2 * root + 1) = joinDeltas(delta(2 * root + 1), delta(root))
    delta(2 * root + 2) = joinDeltas(delta(2 * root + 2), delta(root))
    delta(root) = getNeutralDelta
  }

  def query(from: Int, to: Int): Int = query(from, to, 0, 0, n - 1)

  def query(from: Int, to: Int, root: Int, left: Int, right: Int): Int = {
    if (from == left && to == right) return joinValueWithDelta(value(root), deltaEffectOnSegment(delta(root), right - left + 1))
    pushDelta(root, left, right)
    val mid: Int = (left + right) >> 1
    if (from <= mid && to > mid) queryOperation(query(from, math.min(to, mid), root * 2 + 1, left, mid), query(math.max(from, mid + 1), to, root * 2 + 2, mid + 1, right))
    else if (from <= mid) query(from, math.min(to, mid), root * 2 + 1, left, mid)
    else if (to > mid) query(math.max(from, mid + 1), to, root * 2 + 2, mid + 1, right)
    else throw new RuntimeException("Incorrect query from " + from + " to " + to)
  }

  def modify(from: Int, to: Int, delta: Int) {
    modify(from, to, delta, 0, 0, n - 1)
  }

  def modify(from: Int, to: Int, delta: Int, root: Int, left: Int, right: Int) {
    if (from == left && to == right) {
      this.delta(root) = joinDeltas(this.delta(root), delta)
      return
    }
    pushDelta(root, left, right)
    val mid: Int = (left + right) >> 1
    if (from <= mid) modify(from, math.min(to, mid), delta, 2 * root + 1, left, mid)
    if (to > mid) modify(math.max(from, mid + 1), to, delta, 2 * root + 2, mid + 1, right)
    value(root) = queryOperation(joinValueWithDelta(value(2 * root + 1), deltaEffectOnSegment(this.delta(2 * root + 1), mid - left + 1)), joinValueWithDelta(value(2 * root + 2), deltaEffectOnSegment(this.delta(2 * root + 2), right - mid)))
  }
}

object SegmentTree {
  // Usage example
  def main(args: Array[String]) {
    val rnd = new Random(1)
    for (step <- 0 until 1000) {
      val n = rnd.nextInt(50) + 1
      val t = new SegmentTree(n)
      val x = Array.fill(n)(t.getInitValue)
      for (i <- 0 until 1000) {
        val b = rnd.nextInt(n)
        val a = rnd.nextInt(b + 1)
        val cmd = rnd.nextInt(3)
        if (cmd == 0) {
          val delta = rnd.nextInt(100) - 50
          t.modify(a, b, delta)
          for (j <- a to b)
            x(j) = t.joinValueWithDelta(x(j), delta)
        } else if (cmd == 1) {
          val res1 = t.query(a, b)
          var res2 = x(a)
          for (j <- a to b)
            res2 = t.queryOperation(res2, x(j))
          if (res1 != res2) throw new RuntimeException
        } else {
          for (j <- 0 until n)
            if (t.query(j, j) != x(j))
              throw new RuntimeException
        }
      }
    }
    println("Test passed")
  }
}
